### DISCUSSION TITLE
A question about the load balance in MOOSE

### DISCUSSION THREAD

**Discussion by alimostafavi24** [idaholab/moose#28638] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear MOOSE Experts,

I have a question regarding dynamic load balancing between processors, particularly when using Adaptive Mesh Refinement (AMR). From my understanding, the code tries to partition the domain to equally distribute finite element degrees of freedom across processors. Could someone elaborate on how this load balancing is achieved, and possibly provide references or documentation that further describe the process? I am interested to know the algorithm behind this.

I’ve attached an example from my simulation illustrating the load balancing behavior.

Thank you in advance for your help!

Best regards,

<img width="1161" alt="Screenshot 2024-09-16 at 3 35 02 PM" src="https://github.com/user-attachments/assets/9a0cfc78-506e-4d29-844f-c8f485c03c0c">

**Comment by GiudGiud** [comment]
Hello

The process first depends on your partitioner. Which one are you using?

Guillaume

**Reply by alimostafavi24** [reply]
I use the default partitioner which should be METIS.

**Reply by GiudGiud** [reply]
depends on whether you use distributed mesh
```
      if (use_distributed_mesh)
        partitioner = "parmetis";
      else
        partitioner = "metis";
```

anyway once you have a partitioner, this partitions using a heuristic and more or less of an internal solve to get an optimal partition.
This is repeated on every re-partition. There is communication performed by libmesh (for the mesh), petsc (for the solution vectors) and moose (for stateful material properties)

**Reply by GiudGiud** [reply]
depends on whether you use distributed mesh
```
      if (use_distributed_mesh)
        partitioner = "parmetis";
      else
        partitioner = "metis";
```

anyway once you have a partitioner, this partitions using a heuristic and more or less of an internal solve to get an optimal partition.
This is repeated on every re-partition. There is communication performed by libmesh (for the mesh), petsc (for the solution vectors) and moose (for stateful material properties)

**Reply by alimostafavi24** [reply]
Thanks. It seems like a complicated process that involve MOOSE, libmesh and petsc. I need to cite a paper for dynamic load balancing.  Can you think of any reference from MOOSE, libmesh or petsc that go through the details of the load balance?

**Reply by GiudGiud** [reply]
If you are focused on the load balance part, then it s the partitioners you want to cite. MOOSE/libmesh/petsc re-partition, but the "balance" is provided by the partitioner

**Reply by alimostafavi24** [reply]
Is there any part of the code I can look into to learn about this? "this partitions using a heuristic and more or less of an internal solve to get an optimal partition."

**Reply by GiudGiud** [reply]
you can open one of the partitioner source files, then use vscode "go to reference" to see how its routines are called
